packages/alfa-dom/src/node/element.ts

```diff
@@ -447,6 +448,67 @@ export namespace Element {
     });
   }
 
+  /**
+   * @internal
+   */
+  export function cloneElement(
+    options: Node.ElementReplacementOptions,
+    device?: Device,
+  ): (element: Element) => Trampoline<Element> {
+    return (element) =>
+      Trampoline.traverse(element.children(), (child) => {
+        if (Element.isElement(child) && options.predicate(child)) {
+          return Trampoline.done(Array.from(options.newElements));
+        }
+
+        return Node.cloneNode(child, options, device).map((node) => [node]);
+      }).map((children) => {
+        const deviceOption = Option.from(device);
+        const clonedElement = Element.of(
+          element.namespace,
+          element.prefix,
+          element.name,
+          element.attributes.map((attribute) => Attribute.clone(attribute)),
+          Iterable.flatten(children),
+          element.style.map((block) => {
+            return Block.of(
+              Iterable.map(block.declarations, (declaration) =>
+                Declaration.of(
+                  declaration.name,
+                  declaration.value,
+                  declaration.important,
+                ),
+              ),
+            );
+          }),
+          deviceOption.flatMap((d) => element.getBoundingBox(d)),
+          deviceOption,
+          element.externalId,
+        );
+
+        if (element.shadow.isSome()) {
+          const shadow = element.shadow.get();
+          clonedElement._attachShadow(
+            Shadow.of(
+              shadow
+                .children()
+                .map((child) => Node.clone(child, options, device)),
+              shadow.style,
+              shadow.mode,
+              shadow.externalId,
+              shadow.extraData,
+            ),
```

Comment:
**Suggestion:** we should probably have a `Shadow.clone` instead of inlining it here.